app/models/user_bookmark_list.rb

```diff
@@ -21,10 +21,42 @@ def initialize(user:, guardian:, params:)
 
   def load
     @bookmarks = BookmarkQuery.new(user: @user, guardian: @guardian, params: @params).list_all
+    if SiteSetting.use_polymorphic_bookmarks
+      preload_polymorphic_associations
+    end
     @bookmarks
   end
 
   def per_page
     @per_page ||= @params[:per_page]
   end
+
+  private
+
+  # We have already confirmed the user has access to these records at
+  # this point in BookmarkQuery, so it is safe to load them directly
+  # without any further security checks.
+  #
+  # These polymorphic associations are loaded to make the UserBookmarkListSerializer's
+  # life easier, which conditionally chooses the bookmark serializer to use based
+  # on the type, and we want the associations all loaded ahead of time to make
+  # sure we are not doing N1s.
+  def preload_polymorphic_associations
+    @topics = Topic.includes(:topic_users).where(
+      id: Bookmark.select_type(@bookmarks, "Topic").map(&:bookmarkable_id)
```

Comment:
```suggestion
      id: Bookmark.select_type(@bookmarks, "Topic").select(:bookmarkable_id)
```

If we confirm that `Bookmark.select_type(@bookmarks, "Topic")` is returning a relation, we can just select the column and AR will covert it to a subquery instead of doing a query to fetch the ids.

Reply:
See my comment on bookmark.rb, this causes more queries, and I still have to do the `.map` on the end anyway to get the IDs I need

Reply:
See https://github.com/discourse/discourse/pull/16335#discussion_r847078177